src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs

```diff
@@ -155,6 +155,19 @@ private async Task StartHostAsync(CancellationToken cancellationToken, int attem
             finally
             {
                 _hostStartSemaphore.Release();
+                await RestartIfHostOfflineChanged();
+            }
+        }
+
+        /// <summary>
+        /// If the Host's status is Offline while app_offline.htm does not exist, we need to restart.
+        /// </summary>
+        private async Task RestartIfHostOfflineChanged()
+        {
+            bool isOffline = Utility.CheckAppOffline(_applicationHostOptions.CurrentValue.ScriptPath);
+            if (!isOffline && State == ScriptHostState.Offline)
```

Comment:
I am only doing this when `State` is offline. 

I wonder if there could be a case where the `app_offline.htm` file is there but Host is not offline. Maybe I should take care of that too.

Reply:
Right - both states can happen. I'd just move this check back to where the first check is (in UnsynchronizedStartHostAsync). Then at the end of startup, compare the initial isOffline with the current one. If not equal, restart.

```csharp
bool isOffline = IsOffline();
// startup logic
if (isOffline != IsOffline())
{
  // restart
}
```

Reply:
Yeah I thought about that. Problem is that if we need to restart, we have to release the `_hostStartSemaphore` first to allow another restart. I can do that there as part of the restart. To avoid accidentally releasing the semaphore twice and figured it would be better to let it complete and then check.